lib/bolt/config.rb

```diff
@@ -141,6 +157,11 @@ def initialize(boltdir, config_data, overrides = {})
         @transports[key] = transport.default_options
       end
 
+      defaults.each do |default|
+        @config_files << default[:filepath]
+        update_from_file(default[:data])
```

Comment:
It might make sense to do the merging at the hash level and then call `update_from_file` once rather than calling for each level. That way we could have a single place that just implements the merge behavior. I'm not confident on that one though.

Reply:
I'm not sure either. I mostly based this off of Alex's comment: https://github.com/puppetlabs/bolt/pull/1484#issuecomment-565624774

Reply:
If we merged all the data together, we could also have the 0th layer be the defaults. Then we could avoid all the `if data.key?` checks when assigning the instance variables. It might be worth a spike at least, to see if it simplifies things.

I was imagining something like:

```rb
merged_config = {
  # ... defaults here
}

config_datas.each do |data|
  merged_config.merge!(data) do |key, val1, val2|
    case key
    when <something we want to shallow or deep merge>
      # do the merge
    else
      # use the newer value
      val2
    end
  end
end
```

And then it will be very clear how data is merged.

Reply:
We could then even transform the "overrides" into a hash and merge that on top too, if it uses the same merge strategy.

We might also be able to get rid of a bunch of instance variables if we just make the accessor methods read from the merged config hash.

Reply:
Updated to merge the config hashes prior to calling `update_from_file`. It's definitely possible to have a default hash instead of the instance variables, but it'll take a bit more work that I think could be part of a larger effort to refactor the class.